spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/AbstractJpaQuery.java

```diff
@@ -378,6 +403,14 @@ public boolean containsKey(Object key) {
 					tuple.get((String) key);
 					return true;
 				} catch (IllegalArgumentException e) {
+					if (nativeQuery) {
+						try {
+							tuple.get(JdbcUtils.convertPropertyNameToUnderscoreName((String) key));
```

Comment:
Instead of customizing the `Map`, it would make sense to have a custom `Tuple` wrapper that handles the conversion. 

Consider a tuple that would contain `first_name` and `firstName` elements. It is impossible to tell in which order these should appear in a map nor which value to return in a `Set<Map.Entry>`.

Attempting snake-case to camel-case conversion on value absence makes sense, for the rest we should keep the functionality as-is.

The `Tuple` wrapper would delegate all enumerative calls to the `Tuple` delegate while `get(…)` calls are augmented for property name conversion.

Reply:
Good point, updated now, please review again.